multicorn_das/__init__.py

```
@@ -752,3 +744,112 @@ def multicorn_sortkeys_to_grpc_sortkeys(sortkeys):
+
+    def _rows_generator(self):
+        """
+        A generator that iterates over chunks in the stream, then yields each row.
+        Exits gracefully if we detect that the iterator has been closed.
+        """
+        try:
+            for chunk in self._stream:
+                # If user called .close(), exit early
+                if self._closed:
+                    break
+
+                # Each chunk has multiple rows
+                for row in chunk.rows:
+                    yield build_row(row.data.items())
+
+        except GeneratorExit:
+            # The generator got force-closed (e.g., if someone forcibly closed
+            # the Python generator). We'll just let it end.
+            pass
+
+    def __iter__(self):
+        """
+        Ensure this object is iterable. If we haven't created the generator,
+        do so now.
+        """
+        if self._iterator is None:
+            self._iterator = self._rows_generator()
+        return self
+
+    def __next__(self):
+        """
+        Get the next row from the generator, or close if we've reached the end.
+        """
+        if self._closed:
+            # If .close() was called, we don't return anything further.
+            raise StopIteration("Iterator already closed.")
+
+        try:
+            return next(self._iterator)
+        except StopIteration:
+            # Natural end of the generator => close and re-raise
+            self.close()
+            raise
+        except Exception:
+            # Any other error (including gRPC errors) => ensure we close, then re-raise
+            self.close()
+            raise
+
+    def close(self):
+        """
+        Close the iterator and cancel the gRPC stream (if still open).
+        """
+        if not self._closed:
+            self._closed = True
+            log_to_postgres(
+                f"Cancelling gRPC stream for table {self._table_id}",
+                WARNING
+            )
+            self._stream.cancel()
```

Don't we need to try/catch and ignore any error here?